platform/logger.cpp

```diff
@@ -0,0 +1,116 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(MBED_CONF_RTOS_PRESENT) || !defined(MBED_CONF_ZERO_BUFFER_LOGGING)
+
+#include <string.h>
+#include "platform/mbed_logger.h"
+#include "platform/CircularBuffer.h"
+#include "platform/mbed_wait_api.h"
+#include "hal/ticker_api.h"
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_critical.h"
+
+using namespace mbed;
+
+#if defined (MBED_ID_BASED_TRACING)
+CircularBuffer <uint32_t, (MBED_CONF_LOG_MAX_BUFFER_SIZE/4)> log_buffer;
+#else
+CircularBuffer <char, MBED_CONF_LOG_MAX_BUFFER_SIZE> log_buffer;
+#endif
+const ticker_data_t *const log_ticker = get_us_ticker_data();
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined (MBED_ID_BASED_TRACING)
+void log_thread(void)
+{
+    uint32_t data = 0;
+    while (1) {
+        while (!log_buffer.empty()) {
+            log_buffer.pop(data);
+#if DEVICE_STDIO_MESSAGES && !defined(NDEBUG)
+            printf("0x%x ", data);
+#endif
+        }
+        wait_ms(1);
+    }
+}
+
+// uint32_t time | uint32 (ID) | uint32 args ... | uint32_t checksum | 0
+void log_buffer_id_data(uint8_t argCount, ...)
+{
+    volatile uint64_t time = ticker_read_us(log_ticker);
+    uint32_t data = 0;
+    uint32_t checksum = 0;
+
+    data = (uint32_t)(time/1000);
+    log_buffer.push(data);
+    checksum ^= data;
+
+    va_list args;
+    va_start(args, argCount);
+    for (uint8_t i = 0; i < argCount; i++) {
+        data = va_arg(args, uint32_t);
+        log_buffer.push(data);
+        checksum ^= data;
+    }
+    log_buffer.push(checksum);
+    log_buffer.push(0x0);
+    va_end(args);
+}
+
+#else
+void log_thread(void)
+{
+    char data;
+    while (1) {
+        while (!log_buffer.empty()) {
+            log_buffer.pop(data);
+#if DEVICE_STDIO_MESSAGES && !defined(NDEBUG)
+            printf("%c", data);
+#endif
+        }
+        wait_ms(1);
+    }
+}
+
+void log_buffer_string_data(const char *format, ...)
+{
+    volatile uint64_t time = ticker_read_us(log_ticker);
+    char one_line[MBED_CONF_MAX_LOG_STR_SIZE];
+    uint8_t count = snprintf(one_line, MBED_CONF_MAX_LOG_STR_SIZE, "%-8lld ", time);
+    uint8_t bytes_written = 0;
+
+    va_list args;
+    va_start(args, format);
+    vsnprintf(one_line+count, (MBED_CONF_MAX_LOG_STR_SIZE-count), format, args);
+    count = strlen(one_line);
+    while (count) {
+        log_buffer.push(one_line[bytes_written++]);
```

Comment:
I'm not convinced there's a point to having a manual buffer here. Can we not assume the console has some buffering? (eg by setting `platform.stdio-buffered-serial` on).

This also will drop log output, rather than wait for buffer draining. Do people really want that behaviour? Should certainly be configurable at least.

I guess the buffer allows that behaviour, because C library output assumes blocking? Having a whole thread to do the pumping seems excessive though.

With #5571 just in, I can envisage a sort of setup to achieve it like:

     FileHandle *stderr_fh = get_fhc(STDERR_FILENO); // needs to be exposed, as discussed
     stderr_fh->set_blocking(false);
     stderr_fh->write(one_line, strlen(one_line));
     stderr_fh->set_blocking(true);

The catch being that stderr's buffer may not be big enough for your purpose - it would need to be configured elsewhere to be enough. Unfortunately there's only one setting for all `UARTSerial`, can't be set per-device.

Reply:
Buffering is not to cover up for console/UART buffering. This buffer serves multiple purpose:
1. Can print from interrupts
2. In production software, this buffer can be used to dump last few debug traces in case of system crash.
3. In case of time sensitive debugging, buffer with ID based method will be useful.

Reply:
> This also will drop log output, rather than wait for buffer draining. Do people really want that behaviour? Should certainly be configurable at least.

Buffering is configurable, MBED_CONF_ZERO_BUFFER_LOGGING